src/pybind/mgr/dashboard/controllers/__init__.py

```diff
@@ -454,146 +541,107 @@ class RESTController(BaseController):
-                    (val._collection_method_, attr, '_handle_detail_method', []))
-
-        for attr, val in inspect.getmembers(cls, predicate=isfunction):
-            if hasattr(val, '_resource_method_'):
-                res_params = [":{}".format(arg) for arg in args]
-                url_suffix = "{}/{}".format("/".join(res_params), attr)
-                result.append(
-                    (val._resource_method_, url_suffix, '_handle_detail_method', []))
-
-        return result
+            func = getattr(cls, k, None)
+            while hasattr(func, "__wrapped__"):
+                func = func.__wrapped__
+            if v['resource'] and func:
+                path_params = cls.get_path_param_names()
+                params = _get_function_params(func)
+                return [p['name'] for p in params
+                        if p['required'] and p['name'] not in path_params]
+        return None
 
-    @cherrypy.expose
-    def _collection(self, *vpath, **params):
-        return self._rest_request(False, *vpath, **params)
-
-    @cherrypy.expose
-    def _element(self, *vpath, **params):
-        return self._rest_request(True, *vpath, **params)
-
-    def _handle_detail_method(self, *vpath, **params):
-        method = getattr(self, cherrypy.request.path_info.split('/')[-1])
-
-        if cherrypy.request.method not in ['GET', 'DELETE']:
-            method = RESTController._takes_json(method)
-
-        method = RESTController._returns_json(method)
-
-        cherrypy.response.status = 200
-
-        return method(*vpath, **params)
-
-    def _rest_request(self, is_element, *vpath, **params):
-        method_name, status_code = self._method_mapping[
-            (cherrypy.request.method, is_element)]
-        method = getattr(self, method_name, None)
-
-        if cherrypy.request.method not in ['GET', 'DELETE']:
-            method = RESTController._takes_json(method)
-
-        method = RESTController._returns_json(method)
-
-        cherrypy.response.status = status_code
-
-        return method(*vpath, **params)
-
-    @staticmethod
-    def _function_args(func):
-        return getargspec(func).args[1:]
-
-    @staticmethod
-    def _takes_json(func):
-        def inner(*args, **kwargs):
-            if cherrypy.request.headers.get('Content-Type',
-                                            '') == 'application/x-www-form-urlencoded':
-                return func(*args, **kwargs)
+    @classmethod
+    def endpoints(cls):
+        result = super(RESTController, cls).endpoints()
+        for _, val in inspect.getmembers(cls, predicate=callable):
+            if val.__name__ in cls._method_mapping:
+                meth = cls._method_mapping[val.__name__]
+
+                path = ""
+                if meth['resource']:
+                    res_id_params = cls.infer_resource_id()
+                    if res_id_params is None:
+                        raise TypeError("Invalid resource method '{}'. Please "
```

Comment:
- The structure of the error message is a bit confusing.
- The string format parameter is missing, ``{}`` will be replaced by an empty string otherwise.